File: src/test/java/no/nav/syfo/oppgave/JobbTest.java
package no.nav.syfo.oppgave;

import no.nav.syfo.oppgave.Jobb;
import no.nav.syfo.oppgave.Oppgavetype;
import org.junit.Test;

import java.util.stream.Stream;

import static java.util.stream.Stream.of;
import static no.nav.syfo.oppgave.Oppgavetype.OPPFOELGINGSDIALOG_ARKIVER;
import static no.nav.syfo.oppgave.Oppgavetype.OPPFOELGINGSDIALOG_SEND;
import static org.assertj.core.api.Assertions.assertThat;

public class JobbTest {
    @Test(expected = RuntimeException.class)
    public void oppgavetypeManglerOverride() throws Exception {
        Jobb jobb = oppgavetype -> {
        };
        jobb.skalUtforeOppgave(OPPFOELGINGSDIALOG_SEND);
    }

    @Test
    public void oppgavetypeOverrideSkalUtforeOppgaveTrue() throws Exception {
        Jobb jobb = new Jobb() {
            @Override
            public void utfoerOppgave(String id) {
            }

            @Override
            public Oppgavetype oppgavetype() {
                return OPPFOELGINGSDIALOG_SEND;
            }
        };
        assertThat(jobb.skalUtforeOppgave(OPPFOELGINGSDIALOG_SEND)).isTrue();
    }

    @Test
    public void oppgavetyperOverrideSkalUtforeOppgaveFalse() throws Exception {
        Jobb jobb = new Jobb() {
            @Override
            public void utfoerOppgave(String id) {
            }

            @Override
            public Stream<Oppgavetype> oppgavetyper() {
                return of(OPPFOELGINGSDIALOG_ARKIVER);
            }
        };
        assertThat(jobb.skalUtforeOppgave(OPPFOELGINGSDIALOG_SEND)).isFalse();
    }
}

File: src/test/kotlin/no/nav/syfo/oppgave/JobbTest.kt
package no.nav.syfo.oppgave

import org.assertj.core.api.Assertions
import org.junit.Test
import java.util.stream.Stream

class JobbTest {
    @Test(expected = RuntimeException::class)
    fun oppgavetypeManglerOverride() {
        val jobb = Jobb { }
        jobb.skalUtforeOppgave(Oppgavetype.OPPFOELGINGSDIALOG_SEND)
    }

    @Test
    fun oppgavetypeOverrideSkalUtforeOppgaveTrue() {
        val jobb: Jobb = object : Jobb {
            override fun utfoerOppgave(id: String) {}
            override fun oppgavetype(): Oppgavetype {
                return Oppgavetype.OPPFOELGINGSDIALOG_SEND
            }
        }
        Assertions.assertThat(jobb.skalUtforeOppgave(Oppgavetype.OPPFOELGINGSDIALOG_SEND)).isTrue()
    }

    @Test
    fun oppgavetyperOverrideSkalUtforeOppgaveFalse() {
        val jobb: Jobb = object : Jobb {
            override fun utfoerOppgave(id: String) {}
            override fun oppgavetyper(): Stream<Oppgavetype> {
                return Stream.of(Oppgavetype.OPPFOELGINGSDIALOG_ARKIVER)
            }
        }
        Assertions.assertThat(jobb.skalUtforeOppgave(Oppgavetype.OPPFOELGINGSDIALOG_SEND)).isFalse()
    }
}
